File: day25.py
from collections import defaultdict

tape = defaultdict(int)
cursor = 0

state = 'A'

for _ in range(12964419):
    value = tape[cursor]

    if state == 'A':
        if value == 0:
            tape[cursor] = 1
            cursor += 1
            state = 'B'
        elif value == 1:
            tape[cursor] = 0
            cursor += 1
            state = 'F'
        else:
            0/0

    elif state == 'B':
        if value == 0:
            tape[cursor] = 0
            cursor -= 1
            state = 'B'
        elif value == 1:
            tape[cursor] = 1
            cursor -= 1
            state = 'C'
        else:
            0/0

    elif state == 'C':
        if value == 0:
            tape[cursor] = 1
            cursor -= 1
            state = 'D'
        elif value == 1:
            tape[cursor] = 0
            cursor += 1
            state = 'C'
        else:
            0/0

    elif state == 'D':
        if value == 0:
            tape[cursor] = 1
            cursor -= 1
            state = 'E'
        elif value == 1:
            tape[cursor] = 1
            cursor += 1
            state = 'A'
        else:
            0/0

    elif state == 'E':
        if value == 0:
            tape[cursor] = 1
            cursor -= 1
            state = 'F'
        elif value == 1:
            tape[cursor] = 0
            cursor -= 1
            state = 'D'
        else:
            0/0

    elif state == 'F':
        if value == 0:
            tape[cursor] = 1
            cursor += 1
            state = 'A'
        elif value == 1:
            tape[cursor] = 0
            cursor -= 1
            state = 'E'
        else:
            0/0
    else:
        0 / 0


print(sum(tape.values()))

File: 2017/src/day25.rs
use std::collections::HashMap;

#[derive(Copy, Clone)]
enum State {
    A,
    B,
    C,
    D,
    E,
    F,
}

fn main() {
    let mut tape: HashMap<i32, bool> = HashMap::new();
    let mut cursor = 0;
    let mut state = State::A;

    for _ in 0..12964419 {
        let value = tape.entry(cursor).or_insert(false);

        state = match (state, *value) {
            (State::A, false) => { *value =  true; cursor += 1; State::B },
            (State::A,  true) => { *value = false; cursor += 1; State::F },
            (State::B, false) => { *value = false; cursor -= 1; State::B },
            (State::B,  true) => { *value =  true; cursor -= 1; State::C },
            (State::C, false) => { *value =  true; cursor -= 1; State::D },
            (State::C,  true) => { *value = false; cursor += 1; State::C },
            (State::D, false) => { *value =  true; cursor -= 1; State::E },
            (State::D,  true) => { *value =  true; cursor += 1; State::A },
            (State::E, false) => { *value =  true; cursor -= 1; State::F },
            (State::E,  true) => { *value = false; cursor -= 1; State::D },
            (State::F, false) => { *value =  true; cursor += 1; State::A },
            (State::F,  true) => { *value = false; cursor -= 1; State::E },
        }
    }

    let sum: u32 = tape.values().map(|&v| v as u32).sum();
    println!("{}", sum);
}
